File: Code-Java/src/codejava/base/io/IoEntry.java
package codejava.base.io;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.io.RandomAccessFile;
import java.io.StringReader;
import java.io.StringWriter;
import java.nio.file.Files;


public class IoEntry {
  private final static String filename = "FILENAME.txt";
  private final static int size = 10;
  private static byte[] buf = {0x31, 0x32, 0x33, 0x34, 0x35, 0x36, 0x37, 0x38, 0x39, 0x40};
  public static void IoProgram() throws IOException {
    File tmp = File.createTempFile("jms", ".txt");

    File file = new File(filename);
    Files.deleteIfExists(file.toPath());

    tmp.renameTo(file);
    // File open
    FileInputStream fis = new FileInputStream(filename);
    // File write
    FileOutputStream fos = new FileOutputStream(filename);
    fos.write(buf, 0, buf.length);
    fos.flush();
    // File read
    byte [] buf1 = new byte[10];
    int n = fis.read(buf1, 0, buf1.length);
    System.out.println(new String(buf1));
    fis.close();
    fos.close();

    FileWriter fw = new FileWriter(file, true);

    fw.write("This is a ababababababaaabababaabababa!\n");
    fw.write("Second line is also meaningless.\n");
    fw.write("This is the third line.\n");
    fw.flush();
    fw.close();

    char []chars = new char[10];
    FileReader fr = new FileReader(file);
    fr.read(chars, 0, 10);
    fr.read(chars, 0, 10);
    System.out.println(chars);
    fr.close();

    // String Reader, Writer
    StringWriter sw = new StringWriter();
    char [] charSeq = "return 0;".toCharArray();
    char a = ';';
    sw.write(charSeq, 0, 7);
    sw.write("max(Alice.Money, Bob.Money)");
    sw.write(" >1000? me", 1, 5);
    sw.write(a);

    sw.append('\n');
    sw.append("\\\\ Reconstruction! \n");
    // different from write in parameter meaning
    sw.append("uncompleted or fail", 2,11);
    sw.flush();

    String line = sw.toString();
    System.out.print(line);
    sw.close();

    StringReader sr = new StringReader(line);
    sr.mark(1);
    int c = sr.read();
    sr.reset();
    System.out.println((char) c);

    char[] re = new char[7];
    sr.read(re, 0, re.length);
    System.out.print(re);

    int d = sr.read();
    System.out.println((char) d);
    sr.close();

    RandomAccessFile rf = new RandomAccessFile(file, "rw");
    rf.seek(10);
    int e = rf.readByte();
    System.out.println((char) e);
//
//        long pos = rf.getFilePointer();
//
//        // Java Character length is 2 bytes
//        Character wc = 'c';
//        rf.writeChar(wc);
//        rf.seek(pos);
//        Character rc = rf.readChar();
//        System.out.print("\n" + rc + "\n");
//
//        pos = rf.getFilePointer();
//        rf.writeBytes("Hello, World!");
//        rf.seek(pos);
//
//        byte []read_rf = new byte[10];
//        rf.read(read_rf);
//        System.out.print(new String(read_rf));
//
//        String message = String.format("[%d]\t[%10s]", 10, "Hello");
//        rf.write('\n');
//
//        pos = rf.getFilePointer();
//        rf.writeBytes(message);
//        rf.seek(pos);
//
//        String rf_read = rf.readLine();
//        System.out.print('\n'+rf_read);
//        rf.close();
  }
}

File: Code-CSharp/Base/IO/IoEntry.cs
using System.ComponentModel;
using System.Security.Cryptography;
using System.Text;
using System.IO;

namespace CodeCsharp.Base.IO;

public class IoEntry
{
    public static void IoProgram() {
        string filename = "FILENAME.txt";
        byte[] buf = {0x31, 0x32, 0x33, 0x34, 0x35, 0x36, 0x37, 0x38, 0x39, 0x40};
        // tmp File Name
        string tmp = Path.GetTempFileName();

        FileStream tf = File.Create(tmp);
        // Close after create
        // If not, Create will occupy tf and operations of it will fail
        tf.Close();
        
        // delete File if exist
        if(File.Exists(filename)) File.Delete(filename);
        File.Move(tmp, filename);
        
        // File open
        FileStream fs = File.Open(filename, FileMode.Open);
        // File write
        fs.Write(buf, 0, buf.Length);
        fs.Flush();
        fs.Close();
        
        fs = File.Open(filename, FileMode.Open);
        // File read
        byte[] buf1 = new byte[10];
        int n = fs.Read(buf1, 0, buf1.Length);
        Console.Out.WriteLine(Encoding.ASCII.GetChars(buf1));
        fs.Close();

        using (StreamWriter sw = new StreamWriter(filename, true))
        {
            sw.WriteLine("This is a ababababababaaabababaabababa!");
            sw.WriteLine("Second line is also meaningless.");
            sw.WriteLine("This is the third line.");
        }
        
        using (StreamReader sr = new StreamReader(filename))
        {
            char[] c = new char[10];
            sr.Read(c, 0, c.Length);
            sr.Read(c, 0, c.Length);
            Console.WriteLine(c);

            //string line = sr.ReadLine();
            //Console.WriteLine(line);
        }
        
        StringWriter strWriter = new StringWriter();
        char[] charseq = "return 0;".ToCharArray();
        char a = ';';
        strWriter.Write(charseq, 0, 7);
        strWriter.Write("max(Alice.Money, Bob.Money)");
        char[] str1 = " >1000? me".ToCharArray();
        strWriter.Write(str1, 1, 5);
        strWriter.Write(a);
        
        strWriter.Write('\n');
        strWriter.Write("\\\\ Reconstruction! \n");
        char [] str2 = "uncompleted or fail".ToCharArray();
        strWriter.Write(str2, 2, 9);
        strWriter.Flush();
        
        string line = strWriter.ToString();
        Console.WriteLine(line);
        strWriter.Close();
        
        using (StringReader reader = new StringReader(line))
        {
            int c = reader.Peek();
            Console.Out.WriteLine((char )c);

            char[] re = new char[7];
            reader.Read(re, 0, re.Length);
            Console.Write(re);

            int d = reader.Read();
            Console.Out.WriteLine((char )d);
        }
        
        // string dfile = "FILENAME.dat";
        // BinaryReader

        // Random Access 
        fs = File.Open(filename, FileMode.Open);
        fs.Seek(10, SeekOrigin.Begin);
        int e = fs.ReadByte();
        Console.Out.WriteLine((char )e);
    }
}